From b596353e98514994648bafe471f93b855189366e Mon Sep 17 00:00:00 2001
From: Xiaoyu Yao <xyao@apache.org>
Date: Tue, 8 Sep 2015 09:57:36 -0700
Subject: [PATCH 2061/2848] HDFS-9019. Adding informative message to sticky
 bit permission denied exception. Contributed by
 Xiaoyu Yao.

(cherry picked from commit 970daaa5e44d3c09afd46d1c8e923a5096708c44)
(cherry picked from commit b2465690a15ee4257b381a8d186ffa6981ddc537)
(cherry picked from commit d7da70332f4733e84922c3da7c0a79d084fd276b)

Conflicts:
	hadoop-hdfs-project/hadoop-hdfs/CHANGES.txt
	hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSPermissionChecker.java

Change-Id: I0a0d4106316863d5c023ba2d797b3aef6322a933
---
 .../namenode/DefaultAuthorizationProvider.java     |   14 +++++++++-----
 .../apache/hadoop/fs/permission/TestStickyBit.java |    3 +++
 2 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/DefaultAuthorizationProvider.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/DefaultAuthorizationProvider.java
index 5adcece..737fd98 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/DefaultAuthorizationProvider.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/DefaultAuthorizationProvider.java
@@ -21,7 +21,6 @@
 import org.apache.hadoop.classification.InterfaceAudience;
 import org.apache.hadoop.classification.InterfaceStability;
 import org.apache.hadoop.fs.UnresolvedLinkException;
-import org.apache.hadoop.fs.permission.AclEntry;
 import org.apache.hadoop.fs.permission.AclEntryScope;
 import org.apache.hadoop.fs.permission.AclEntryType;
 import org.apache.hadoop.fs.permission.FsAction;
@@ -29,9 +28,7 @@
 import org.apache.hadoop.hdfs.server.namenode.snapshot.Snapshot;
 import org.apache.hadoop.hdfs.util.ReadOnlyList;
 import org.apache.hadoop.security.AccessControlException;
-import org.apache.hadoop.util.StringUtils;
 
-import java.util.List;
 import java.util.Set;
 import java.util.Stack;
 
@@ -386,8 +383,15 @@ private void checkStickyBit(String user, INode parent, INode inode,
       return;
     }
 
-    throw new AccessControlException("Permission denied by sticky bit setting:" 
-        + " user=" + user + ", inode=" + inode);
+    final String path = inode.getFullPathName();
+    throw new AccessControlException(String.format(
+            "Permission denied by sticky bit: user=%s, path=\"%s\":%s:%s:%s%s, " +
+            "parent=\"%s\":%s:%s:%s%s", user,
+            path, inode.getUserName(), inode.getGroupName(),
+            inode.isDirectory() ? "d" : "-", inode.getFsPermission().toString(),
+            path.substring(0, path.length() - inode.toString().length() - 1 ),
+            parent.getUserName(), parent.getGroupName(),
+            parent.isDirectory() ? "d" : "-", parent.getFsPermission().toString()));
   }
 
   /**
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/permission/TestStickyBit.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/permission/TestStickyBit.java
index 9d0e31b..d5cece4 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/permission/TestStickyBit.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/permission/TestStickyBit.java
@@ -140,6 +140,9 @@ private void confirmDeletingFiles(Configuration conf, Path p)
     } catch (IOException ioe) {
       assertTrue(ioe instanceof AccessControlException);
       assertTrue(ioe.getMessage().contains("sticky bit"));
+      assertTrue(ioe.getMessage().contains("user="+user2.getUserName()));
+      assertTrue(ioe.getMessage().contains("path=\"" + file + "\""));
+      assertTrue(ioe.getMessage().contains("parent=\"" + file.getParent() + "\""));
     }
   }
 
-- 
1.7.9.5

